File: src/cpp-sample/print-detailed-info.cpp
#include <xt-cpp.hpp>
#include <iostream>

static void OnFatal() {
  std::cout << "Fatal error.\n";
}

static void OnTrace(Xt::Level level, const std::string& message) {
  if (level != Xt::Level::Info)
    std::cout << "-- " << level << ": " << message << "\n";
}

int PrintInfoMain(int argc, char** argv) {

  try {
    Xt::Audio init("Sample", nullptr, &OnTrace, &OnFatal);
    std::cout << "Win32: " << Xt::Audio::IsWin32() << "\n";
    std::cout << "Version: " << Xt::Audio::GetVersion() << "\n";
    std::cout << "Pro Audio: " << Xt::Audio::GetServiceBySetup(Xt::Setup::ProAudio)->GetName() << "\n";
    std::cout << "System Audio: " << Xt::Audio::GetServiceBySetup(Xt::Setup::SystemAudio)->GetName() << "\n";
    std::cout << "Consumer Audio: " << Xt::Audio::GetServiceBySetup(Xt::Setup::ConsumerAudio)->GetName() << "\n";

    for (int32_t s = 0; s < Xt::Audio::GetServiceCount(); s++) {
      std::unique_ptr<Xt::Service> service = Xt::Audio::GetServiceByIndex(s);
      std::cout << "Service " << service->GetName() << ":\n";
      std::cout << "  System: " << service->GetSystem() << "\n";
      std::cout << "  Device count: " << service->GetDeviceCount() << "\n";
      std::cout << "  Capabilities: " << service->GetCapabilities() << "\n";
      std::unique_ptr<Xt::Device> defaultInput = service->OpenDefaultDevice(false);
      if(defaultInput)
        std::cout << "  Default input: " << *defaultInput << "\n";
      std::unique_ptr<Xt::Device> defaultOutput = service->OpenDefaultDevice(true);
      if(defaultOutput)
        std::cout << "  Default output: " << *defaultOutput << "\n";

      for (int d = 0; d < service->GetDeviceCount(); d++) {
        std::unique_ptr<Xt::Device> device = service->OpenDevice(d);
        std::cout << "  Device " << device->GetName() << ":\n";
        std::cout << "    System: " << device->GetSystem() << "\n";
        std::unique_ptr<Xt::Mix> mix = device->GetMix();
        if(mix)
          std::cout << "    Current mix: " << *mix << "\n";
        else
          std::cout << "    Current mix: null\n";
        std::cout << "    Input channels: " << device->GetChannelCount(false) << "\n";
        std::cout << "    Output channels: " << device->GetChannelCount(true) << "\n";
        std::cout << "    Interleaved access: " << device->SupportsAccess(true) << "\n";
        std::cout << "    Non-interleaved access: " << device->SupportsAccess(false) << "\n";
      }
    }
  } catch(const Xt::Exception& e) {
    std::cout << "Error: " 
              << "system " << Xt::Exception::GetSystem(e.GetError()) << ", "
              << "fault " << Xt::Exception::GetFault(e.GetError()) << ", "
              << "cause " << Xt::Exception::GetCause(e.GetError()) << ", "
              << "text " << Xt::Exception::GetText(e.GetError()) << ", "
              << "message " << e << ".\n";
  }
  return 0;
}

File: src/cli-sample/PrintDetailedInfo.cs
using System;

namespace Xt {

    public static class PrintDetailedInfo {

        static void OnFatal() {
            Console.WriteLine("Fatal error.");
        }

        static void OnTrace(XtLevel level, string message) {
            if (level != XtLevel.Info)
                Console.WriteLine("-- " + level + ": " + message);
        }

        [STAThread]
        public static void Main(string[] args) {

            try {
                using (XtAudio audio = new XtAudio("Sample", IntPtr.Zero, OnTrace, OnFatal)) {
                    Console.WriteLine("Win32: " + XtAudio.IsWin32());
                    Console.WriteLine("Version: " + XtAudio.GetVersion());
                    Console.WriteLine("Pro Audio: " + XtAudio.GetServiceBySetup(XtSetup.ProAudio).GetName());
                    Console.WriteLine("System Audio: " + XtAudio.GetServiceBySetup(XtSetup.SystemAudio).GetName());
                    Console.WriteLine("Consumer Audio: " + XtAudio.GetServiceBySetup(XtSetup.ConsumerAudio).GetName());

                    for (int s = 0; s < XtAudio.GetServiceCount(); s++) {
                        XtService service = XtAudio.GetServiceByIndex(s);
                        Console.WriteLine("Service " + service.GetName() + ":");
                        Console.WriteLine("  System: " + service.GetSystem());
                        Console.WriteLine("  Device count: " + service.GetDeviceCount());
                        Console.WriteLine("  Capabilities: " + service.GetCapabilities());

                        using (XtDevice defaultInput = service.OpenDefaultDevice(false)) {
                            Console.WriteLine("  Default input: " + defaultInput);
                        }
                        using (XtDevice defaultOutput = service.OpenDefaultDevice(true)) {
                            Console.WriteLine("  Default output: " + defaultOutput);
                        }

                        for (int d = 0; d < service.GetDeviceCount(); d++)
                            using (XtDevice device = service.OpenDevice(d)) {
                                Console.WriteLine("  Device " + device.GetName() + ":");
                                Console.WriteLine("    System: " + device.GetSystem());
                                Console.WriteLine("    Current mix: " + device.GetMix());
                                Console.WriteLine("    Input channels: " + device.GetChannelCount(false));
                                Console.WriteLine("    Output channels: " + device.GetChannelCount(true));
                                Console.WriteLine("    Interleaved access: " + device.SupportsAccess(true));
                                Console.WriteLine("    Non-interleaved access: " + device.SupportsAccess(false));
                            }
                    }
                }
            } catch (XtException e) {
                Console.WriteLine(string.Format("Error: system {0}, fault {1}, cause {2}, text {3}, message: {4}.\n",
                    XtException.GetSystem(e.GetError()),
                    XtException.GetFault(e.GetError()),
                    XtException.GetCause(e.GetError()),
                    XtException.GetText(e.GetError()),
                    e.ToString()));
            }
        }
    }
}
